### PR TITLE
fix: do not strip dotted components from lean module names

### LINKED ISSUE #2999: Imports not working for module names with dots
### Prerequisites

* [x] Put an X between the brackets on this line if you have done all of the following:
    * Check that your issue is not already [filed](https://github.com/leanprover/lean4/issues).
    * Reduce the issue to a minimal, self-contained, reproducible test case. Avoid dependencies to mathlib4 or std4.

### Description

To import a file `A/B.C.lean`, `import «A».«B.C»` instead tries to find a file called `A/B.lean`.

The workaround was to use `import «A».«B.C.lean»` instead.

### Context

Bug described [here](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/Import.20file.20with.20multiple.20dots.20in.20file.20name)

### Steps to Reproduce

1. Create a file `A/B.C.lean`
2. Try to import it from a different file with `import «A».«B.C»`

**Expected behavior:** Import works

**Actual behavior:**

```lean4
`/home/vscode/.elan/toolchains/leanprover--lean4---v4.2.0/bin/lake print-paths Init «A».«B.C» --no-build` failed:

stderr:
error: '«A».«B.C»': no such file or directory (error code: 2)
  file: ./././A/B.lean
```

### Versions

Lean v4.2.0

### Additional Information

[Additional information, configuration or data that might be necessary to reproduce the issue]

### Impact

Add :+1: to [issues you consider important](https://github.com/leanprover/lean4/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc). If others are impacted by this issue, please ask them to add :+1: to it.

### PR DESCRIPTION
**Pull request by eric-wieser**
This introduces `FilePath.addExtension` to take a path that we know has no prior extension, and append a new extension to it.
As this function is simpler than `FilePath.withExtension`, this change eagerly replaces uses of the latter with the former, except in a few cases where stripping the extension really is the right thing to do.

This should fix the bug described at https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/Import.20file.20with.20multiple.20dots.20in.20file.20name/near/404508048, where `import «A.B».«C.D.lean»` is needed to import `A.B/C.D.lean`.

Closes #2999

### CODE REVIEW

File: src/lake/tests/init/test.sh [hunk 1]
```diff
@@ -77,6 +77,12 @@ $LAKE -d lean-data exe lean-data
 $LAKE new 123-hello
 $LAKE -d 123-hello exe 123-hello
 
+# Test creating packages both starting with a digit and containing a `.`
+# https://github.com/leanprover/lean4/issues/2999
+
+$LAKE new 123-hello.world
+$LAKE -d 123-hello-world exe 123-hello.world
+
```
Comment (tydeu) [inline src/lake/tests/init/test.sh:85]: I am pretty sure this does not test what you think it does. This creates a package with a module named `123Hello.World` located `123Hello/World.lean`. If you want to include the do in the module name, you need to use guillemets (e.g., `lake new «A.B».«C.D»`).

Comment (eric-wieser) [reply src/lake/tests/init/test.sh:85]: That creates a folder name that contains the guillemets!

Comment (tydeu) [reply src/lake/tests/init/test.sh:85]: @eric-wieser Oh, yeah. The guillemets in top-level directory name should be stripped (in another PR). Nonetheless, that should create the proper directory structure underneath it (i.e., ` «A-B»-«C-D»/A.B/C.D.lean`) with the fix in this PR.

File: src/lake/tests/init/clean.sh [hunk 2]
```diff
@@ -5,5 +5,6 @@ rm -rf hello_world
 rm -rf hello-exe
 rm -rf lean-data
 rm -rf 123-hello
+rm -rf «A.B».«C.D»
```
Comment (tydeu) [inline src/lake/tests/init/clean.sh:8]:
```suggestion
rm -rf «A-B»-«C-D»
```

**Review by tydeu — APPROVED** [review]
LGTM!

**Comment by tydeu** [comment]
@Kha I think the test failure on Nix may be due to the use of guillemets in a file name?

**Review by mhuisi — APPROVED** [review]
The changes to `FilePath.lean`, `Watchdog.lean` and `Path.lean` look good to me.

**Comment by eric-wieser** [comment]
It looks to me like the full tests are still not actually running

**Comment by Kha** [comment]
@eric-wieser Could you try merging master?

File: src/lake/Lake/CLI/Init.lean [hunk 3]
```diff
@@ -218,16 +218,17 @@ def initPkg (dir : FilePath) (name : String) (tmp : InitTemplate) (env : Lake.En
 
 def validatePkgName (pkgName : String) : LogIO PUnit := do
   if pkgName.isEmpty || pkgName.all (· == '.') || pkgName.any (· ∈ ['/', '\\']) then
-    error "illegal package name"
+    error ("illegal package name: " ++ pkgName)
```
Comment (tydeu) [inline src/lake/Lake/CLI/Init.lean:221]:
```suggestion
    error s!"illegal package name '{pkgName}'"
```

File: src/lake/Lake/CLI/Init.lean [hunk 4]
```diff
@@ -218,16 +218,17 @@ def initPkg (dir : FilePath) (name : String) (tmp : InitTemplate) (env : Lake.En
...
 
 def init (pkgName : String) (tmp : InitTemplate) (env : Lake.Env) (cwd : FilePath := ".") : LogIO PUnit := do
   let pkgName ← do
     if pkgName == "." then
-      match (← IO.FS.realPath cwd).fileName with
+      let path ← IO.FS.realPath cwd
+      match path.fileName with
       | some dirName => pure dirName
-      | none => error "illegal package name"
+      | none => error ("illegal package name: cwd is " ++ path.toString)
```
Comment (tydeu) [inline src/lake/Lake/CLI/Init.lean:231]:
```suggestion
      | none => error s!"illegal package name: could not derive a package name from '{path}'"
```

**Review by tydeu — COMMENTED** [review]
If you intend to keep the error message changes, I suggested some edits I would like to see.

**Comment by tydeu** [comment]
@eric-wieser As to the error itself, I believe it has to do with MSYS2 path conversion. Adding the following to the top of the test file should hopefully fix it:
```shell
# make paths behave on MSYS2
[ "$OSTYPE" == "msys" ] && export MSYS2_ARG_CONV_EXCL=*
```

File: src/lake/tests/init/test.sh [hunk 5]
```diff
@@ -77,6 +77,15 @@ $LAKE -d lean-data exe lean-data
 $LAKE new 123-hello
 $LAKE -d 123-hello exe 123-hello
 
+# Test creating packages with components that contain `.`s
+# https://github.com/leanprover/lean4/issues/2999
+  
+# this fails on windows for unrelated reasons
+if [ "$OSTYPE" != "msys" ]; then
```
Comment (eric-wieser) [inline src/lake/tests/init/test.sh:84]: @tydeu: I couldn't work this one out: but I figure that whatever this bug is, it's not in scope for this PR.

Comment (tydeu) [reply src/lake/tests/init/test.sh:84]: Fair enough. Yeah, I think it can be pushed to a later date.

Comment (eric-wieser) [reply src/lake/tests/init/test.sh:84]: @tydeu: is this PR good to go then?

**Comment by tydeu** [comment]
@eric-wieser I am happy with the Lake changes, but I think we need @Kha to sign off on the core changes before merging.

File: src/Init/System/FilePath.lean [hunk 6]
```diff
@@ -101,6 +101,11 @@ def withFileName (p : FilePath) (fname : String) : FilePath :=
   | none => ⟨fname⟩
   | some p => p / fname
 
+def addExtension (p : FilePath) (ext : String) : FilePath :=
```
Comment (Kha) [inline src/Init/System/FilePath.lean:104]: Needs docstring

**Review by Kha — CHANGES_REQUESTED** [review]

**Comment by eric-wieser** [comment]
@Kha, is the new docstring ok?

File: src/Init/System/FilePath.lean [hunk 7]
```diff
@@ -101,11 +101,21 @@ def withFileName (p : FilePath) (fname : String) : FilePath :=
   | none => ⟨fname⟩
   | some p => p / fname
 
+/-- Append the extension `ext` to a path `p`.
+
+`ext` should not contain a `.`, as this function adds one.
```
Comment (Kha) [inline src/Init/System/FilePath.lean:106]:
```suggestion
/-- Appends the extension `ext` to a path `p`.

`ext` should not contain a leading `.`, as this function adds one.
```

**Pull request merged**